core/extension/gdextension.cpp

```diff
@@ -914,6 +914,35 @@ Error GDExtensionResourceLoader::load_gdextension_resource(const String &p_path,
 		return ERR_INVALID_DATA;
 	}
 
+	// Optionally check maximum compatibility.
+	if (config->has_section_key("configuration", "compatibility_maximum")) {
+		uint32_t compatibility_maximum[3] = { 0, 0, 0 };
+		String compat_string = config->get_value("configuration", "compatibility_maximum");
+		Vector<int> parts = compat_string.split_ints(".");
+		for (int i = 0; i < 3; i++) {
+			if (i < parts.size() && parts[i] >= 0) {
+				compatibility_maximum[i] = parts[i];
+			} else {
+				// If a version part is missing, set the maximum to an arbitrary high value.
+				compatibility_maximum[i] = 9999;
+			}
+		}
+
+		compatible = true;
+		if (VERSION_MAJOR != compatibility_maximum[0]) {
+			compatible = VERSION_MAJOR < compatibility_maximum[0];
+		} else if (VERSION_MINOR != compatibility_maximum[1]) {
+			compatible = VERSION_MINOR < compatibility_maximum[1];
+		} else {
+			compatible = VERSION_PATCH <= compatibility_maximum[2];
+		}
+
+		if (!compatible) {
+			ERR_PRINT(vformat("GDExtension only compatible with Godot version %d.%d.%d or earlier: %s", compatibility_maximum[0], compatibility_maximum[1], compatibility_maximum[2], p_path));
```

Comment:
Should probably do something like `4.1.x` when it's unknown here, as the compatibility maximum is set to `9999` when unspecified, otherwise it'd be quite cryptic I'd say

Reply:
Ah, good point! I've updated it to show the same string as was used in the .gdextension file.